modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformaceStatisticsProcessor.java

```diff
@@ -0,0 +1,264 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.performancestatistics;
+
+import java.io.Serializable;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.IgniteFeatures;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.processors.GridProcessorAdapter;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryType;
+import org.apache.ignite.internal.util.GridIntList;
+import org.apache.ignite.internal.util.distributed.DistributedProcess;
+import org.apache.ignite.internal.util.future.GridFinishedFuture;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.lang.IgniteFuture;
+import org.apache.ignite.lang.IgniteFutureCancelledException;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.spi.discovery.DiscoveryDataBag;
+
+import static org.apache.ignite.internal.GridComponent.DiscoveryDataExchangeType.PERFORMANCE_STAT_PROC;
+import static org.apache.ignite.internal.IgniteFeatures.allNodesSupports;
+import static org.apache.ignite.internal.util.distributed.DistributedProcess.DistributedProcessType.PERFORMANCE_STATISTICS;
+
+/**
+ * Performance statistics processor.
+ * <p>
+ * Manages collecting statistics.
+ */
+public class PerformaceStatisticsProcessor extends GridProcessorAdapter implements IgnitePerformanceStatistics {
+    /** Process to start/stop statistics. */
+    private final DistributedProcess<Boolean, Boolean> proc;
+
+    /** Performance statistics writer. */
+    private final FilePerformanceStatisticsWriter writer;
+
+    /** Synchronization mutex for request futures. */
+    private final Object mux = new Object();
+
+    /** Enable/disable statistics request futures. */
+    private final ConcurrentMap<UUID, GridFutureAdapter<Void>> reqFuts = new ConcurrentHashMap<>();
+
+    /** Disconnected flag. */
+    private volatile boolean disconnected;
+
+    /** Stopped flag. */
+    private volatile boolean stopped;
+
+    /** @param ctx Kernal context. */
+    public PerformaceStatisticsProcessor(GridKernalContext ctx) {
+        super(ctx);
+
+        writer = new FilePerformanceStatisticsWriter(ctx);
+
+        proc = new DistributedProcess<>(ctx, PERFORMANCE_STATISTICS, start -> {
```

Comment:
Why do we need `DistributedProcess` here?
What is the "process" here in the first place?
Let's implement this as 'DistributedMetaStorage' listener.

Reply:
We can use `DistributedBooleanProperty` approach.

Reply:
I have mplemented this as 'DistributedMetaStorage' listener.